src/crimson/os/seastore/btree/fixed_kv_node.h

```diff
@@ -27,29 +28,476 @@ namespace crimson::os::seastore {
 
   virtual ~FixedKVNode() = default;
+  virtual node_key_t get_key_from_idx(uint16_t idx) const = 0;
+
+  template<typename iter_t>
+  void update_child_ptr(iter_t iter, ChildableCachedExtent* child) {
+    children[iter.get_offset()] = child;
+    set_child_ptracker(child);
+  }
+
+  virtual bool is_leaf_and_has_children() const = 0;
+
+  template<typename iter_t>
+  void insert_child_ptr(iter_t iter, ChildableCachedExtent* child) {
+    auto raw_children = children.data();
+    auto offset = iter.get_offset();
+    std::memmove(
+      &raw_children[offset + 1],
+      &raw_children[offset],
+      (get_node_size() - offset) * sizeof(ChildableCachedExtent*));
+    if (child) {
+      children[offset] = child;
+      set_child_ptracker(child);
+    } else {
+      // this can only happen when reserving lba spaces
+      ceph_assert(is_leaf_and_has_children());
+      // this is to avoid mistakenly copying pointers from
+      // copy sources when committing this lba node, because
+      // we rely on pointers' "nullness" to avoid copying
+      // pointers for updated values
+      children[offset] = RESERVATION_PTR;
+    }
+  }
+
+  template<typename iter_t>
+  void remove_child_ptr(iter_t iter) {
+    LOG_PREFIX(FixedKVNode::remove_child_ptr);
+    auto raw_children = children.data();
+    auto offset = iter.get_offset();
+    SUBTRACE(seastore_fixedkv_tree, "trans.{}, pos {}, total size {}, extent {}",
+      this->pending_for_transaction,
+      offset,
+      get_node_size(),
+      (void*)raw_children[offset]);
+    // parent tracker of the child being removed will be
+    // reset when the child is invalidated, so no need to
+    // reset it here
+    std::memmove(
+      &raw_children[offset],
+      &raw_children[offset + 1],
+      (get_node_size() - offset - 1) * sizeof(ChildableCachedExtent*));
+  }
+
+  FixedKVNode& get_stable_for_key(node_key_t key) {
+    ceph_assert(is_pending());
+    if (is_mutation_pending()) {
+      return (FixedKVNode&)*get_prior_instance();
+    } else {
+      ceph_assert(!copy_sources.empty());
+      auto it = copy_sources.upper_bound(key);
+      it--;
+      auto &copy_source = *it;
+      ceph_assert(copy_source->get_node_meta().is_in_range(key));
+      return *copy_source;
+    }
+  }
+
+  static void push_copy_sources(
+    FixedKVNode &dest,
+    FixedKVNode &src)
+  {
+    ceph_assert(dest.is_initial_pending());
+    if (!src.is_pending()) {
+      dest.copy_sources.emplace(&src);
+    } else if (src.is_mutation_pending()) {
+      dest.copy_sources.emplace(
+	src.get_prior_instance()->template cast<FixedKVNode>());
+    } else {
+      ceph_assert(src.is_initial_pending());
+      dest.copy_sources.insert(
+	src.copy_sources.begin(),
+	src.copy_sources.end());
+    }
+  }
+
+  virtual uint16_t get_node_split_pivot() = 0;
+
+  static void move_child_ptrs(
+    FixedKVNode &dest,
+    FixedKVNode &src,
+    size_t dest_start,
+    size_t src_start,
+    size_t src_end)
+  {
+    std::memmove(
```

Comment:
Because dest.children.data() and src.children.data() are non-overlapping, std::memcpy should be safe to use and will be more performant.